README.md
# HTMLBuilder

A tool that allows you to generate HTML content from a template in an optimised way. Indeed, instead of adding new elements in the DOM via `innerHTML` (which is not cool), this class analyses a template & builds the elements with the `createElement()` method. This template is an easy-to-read architecture especially when there is a lot of HTML to generate.

HTMLBuilder is written in TypeScript, however there is its equivalent in JavaScript.

## Get started

Get the source code at `src/HTMLBuilder.js`. Then just add it to your project:

```html
<script src="HTMLBuilder.js"></script>
```

Then initialize an instance of HTMLBuilder.

```javascript
var builder = new HTMLBuilder();
```

## Write a template

Let's say you want to generate a button inside an element without a single line of HTML. The first thing to do is to specify the parent element in which to place the future button. How do you do this?

By default, the parent element is `document.body`. Change that in the constructor:

```javascript
var container = document.querySelector("#container");
var builder = new HTMLBuilder(container);
```

Or use:

```javascript
builder.setParent(container);
```

All right, now let's create the template of our button.

```javascript
// You absolutely need to use these quotes if you have several lines: `...`.
// Indeed, the parser reads line by line with a single element on each line.
builder.generate(`
    button
`);
```

Do you see it? A button magically appeared! However, this button is empty, it has no class, no id, nothing! Let's add those things.

In order to add a class, an id etc., you must respect a specific order:

_tagname-(class)-(id)-(content)-(attributes)-(events)_

This means that the tagname is the starting point and it is required. However, the class, the id, the content, the attributes and the events are not necessary, so you can skip them.

Each part (class, id, etc.) has its specific syntax.

| part       | model                               | example                            |
| ---------- | ----------------------------------- | ---------------------------------- |
| parent     | `<name>`                            | "button"                           |
| class      | `.<name>{0,}`                       | ".class1.class2.class3"            |
| id         | `#<name>{0,1}`                      | "#this-is-the-only-id"             |
| content    | `(<content>){0,1}`                  | "(this is content)"                |
| attributes | `[name1=value1; name2=value2]{0,1}` | "[contenteditable; data-y=[a, b]]" |
| events     | `@event1;event2{0,}`                | "@hoverMe;outOfMe"                 |

NOTE: `{0,}` means 0 ore more times & `{0,1}` means 0 or 1 time.

If you don't respect this syntax, you may get strange results...

For a better understanding, let's make a real example by comparing a real HTML element & its translation with this syntax:

```html
<button class="btn-primary" type="button" id="yeah">Click!</button>
```

```javascript
builder.generate(`
    button.btn-primary#yeah(Click!)[type=button]
`);
```

Adding events is more complex, see: [Add en event listener](#add-an-event-listener).

NOTE: HTML entities (`&amp;`, etc.) are automatically decoded.

## Build an architecture

I want to create a button inside a div. How do I do it? This is a complex situation because we need to write implicitly that one element is the child of another. Moreover, a child can be the child of a child etc.

Fortunately, the algorithm can guess by itself which element is the child of which element. Therefore, all you need to do is specify an indentation level for each line _except the first one_.

```javascript
builder.generate(`
    div.box.bg-black
        >button(Click me!)[type=button]
`);
```

Here is the result in HTML (open the console and see it by yourself):

```html
<div class="box bg-black">
  <button type="button">Click me!</button>
</div>
```

As you can see the `>` explains to the algorithm that this line has an indentation level of 1. Therefore, in order to build more sophisticated architectures, just add more of these `>`:

Example:

```html
<div class="box">
  <h1>Title</h1>
  <div class="menu">
    <ul>
      <li>Item 1</li>
      <li>Item 2</li>
      <li>Item 3</li>
    </ul>
  </div>
  <div class="section">
    <h2>I have actually no idea</h2>
  </div>
</div>
```

The above HTML code is generated thanks to:

```javascript
builder.generate(`
    div.box
        >h1(Title)
        >div.menu
            >>ul
                >>>li(Item 1)
                >>>li(Item 2)
                >>>li(Item 3)
        >div.section
            >>h2(I have actually no idea)
`);
```

Just know that you can add several elements at the same time:

```javascript
builder.generate(`
    div.box
        >h1(Title)
        >div.menu
            >>ul
                >>>li(Item 1)
                >>>li(Item 2)
                >>>li(Item 3)
        >div.section
            >>h2(I have actually no idea)
    div.another-box
        >h1(Maybe you should take a look at the technical part below)
`);
```

## Add an event listener

Once you have your button, you realize that it is completely useless without an event listener. That's what happened in my head, so... I added event listeners!

Before writing a template, you must register a new event listener. Indeed, you register an event, then its name will be used as a reference inside the template, and the algorithm will do its magic :). But how do you register an event then?

```javascript
builder.bindEvent({
  name: "sayHello", // give the listener a name in order to recognize it
  type: "click", // the type of event (click, mouseover, mouseout, etc.)
  callback: function (e) {
    console.log("You clicked me!");
  },
  options: true, // whatever you want to add as options to the event listener...
});
```

NOTE: `options` is just the third argument of the method `addEventListener()`. If you don't know what I'm talking about, check the documentation on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener).

Afterwards, let's write our template:

```javascript
builder.generate(`
    button(Click me!)[type=button]@sayHello
`);
```

Note that you can add multiple event listeners to a single element by separating them with a semicolon:

```javascript
// ...
builder.generate(`
    button(Click me!)[type=button]@sayHello;andHoverMe
`);
```

Perfect, now you're good.

## Some details

You can add several attributes to a single element by separating them with a semicolon `;`. This symbol can be changed with the following method:

```javascript
builder.changeSymbolBetweenAttributes("/");
builder.generate(`
    button(Great button)[type=button / data-for=thing]
`);
```

Since 1.0.3, there is a brand new method (I needed it actually):

```javascript
builder.indentTemplate(`>button[type=button]`, 2);
// adds an indentation level of 2, so the result will be a button with an indentation level of 3
// `>>>button[type=button]`
// This method can be used with several lines: each line will be modified.
```

All of the other methods are private (they start with `_`). Ignore them.

## Technical part: the algorithm

How HTMLBuilder builds elements from a template? There are several steps:

- First, the program reads all the lines in order to identify the main HTML elements, i.e. those without indentation (so they don't start with a `>`). A list of lines is created.

- For each main element of our template, the program will read all the lines from the index of the main line to the index of the next main line in the list of lines. If there is not next main line, then the program will read all the lines to the end of the template. This process allows the algorithm to store all the children of each main element.

- We generate the HTML elements thanks that monster:

```javascript
/(\w+)((?:\.[\w-]*)*)*(#[\w-]*)?(?:\((.*)\))?(?:\[(.*)\])?(?:\@([\w;-]*))*/
```

- We now have a list of children for each main element. We save the indentation level of each child. The challenge here is to identify which element is the parent of which child. Therefore, we need to start with the deepest element and work our way up to the element closest to the parent.

* First, we need determine the last child to have the highest level of indentation: we call it the `deepest element`.

* Next, we need to identify its parent element, so we get the `nearest parent`. The deepest element has as parent the nearest element that has an identation level equal to "child's level - 1".

* Finally, as we read the list of children from bottom to top, we `prepend()` in order to keep the right order.

* At each iteration, a child is removed from the list of children because it is placed inside its parent until there are no more children.

Now we just `append()` each main element inside the given parent (by default: `document.body`).

I hope that was clear. Feel free to take a look at the code, everything is explained.

Enjoy!

## License

MIT License


CHANGELOG.md
# [1.0.6] - April 11, 2021

+ Bug fixes.

# [1.0.5] - March 6, 2021

+ Minor fixes.

# [1.0.4] - March 5, 2021

+ Improve regular expression.

# [1.0.3] - March 5, 2021

+ Add method indentTemplate() in order to concatenate templates.
+ Improve documentation.

# [1.0.0] - March 1st, 2021

Initial release


src/HTMLBuilder.ts
// FRENCH="àèìòùÀÈÌÒÙáéíóúýÁÉÍÓÚÝâêîôûÂÊÎÔÛãñõÃÑÕäëïöüÿÄËÏÖÜŸçÇßØøÅåÆæœ"

interface Listener {
	name: string;
	type: string;
	callback: (e: any) => void;
	options?: any;
}

/**
 * A tool that allows you to generate HTML content from a template in an optimised way.
 * @class
 */
class HTMLBuilder {
	/**
	 * The regular expression used to parse a template.
	 * @type {RegExp}
	 * @constant
	 * @private
	 */
	private REGEX: RegExp = /(\w+)((?:\.[\w-]*)*)*(#[\w-]*)?(?:\((.*)\))?(?:\[(.*)\])?(?:\@([\w;-]*))*/;

	/**
	 * The parent element in which to put the generated elements from the template.
	 * @type {HTMLElement}
	 * @private
	 */
	private parent: HTMLElement;

	/**
	 * The symbol uses to separate different attributes.
	 * @type {string}
	 * @private
	 */
	private SYMBOL_BETWEEN_ATTRIBUTES: string = ";";

	/**
	 * The list of all the events.
	 * @private
	 */
	private EVENTS: Listener[] = [];

	/**
	 * @constructs HTMLBuilder
	 * @param {HTMLElement} parent The parent in which to put the generated elements.
	 */
	public constructor(parent?: HTMLElement) {
		this.parent = parent || document.body;
	}

	/**
	 * Changes the parent element.
	 * @param parent The new parent element in which to put the generated elements.
	 * @public
	 */
	public setParent(parent: HTMLElement): void {
		this.parent = parent;
	}

	/**
	 * Registers an event to use in a template. Those events are available for all the templates.
	 * @param {{name: string, type: string, callback: Function, options: any}} event The event to register.
	 * @public
	 */
	public bindEvent(event: Listener): void {
		if (!event.name) throw new Error("bindEvent(): cannot bind an event without a name.");
		if (!event.type) throw new Error("bindEvent(): cannot bind an event without a precise type.");
		if (!event.callback) throw new Error("bindEvent(): cannot bind an event without a callback function.");
		if (event.name.startsWith("on")) {
			event.name = event.name.replace("on", "");
		}
		this.EVENTS.push(event);
	}

	/**
	 * Clears all the events.
	 * @public
	 * @since 1.0.6
	 */
	public clearEvents(): void {
		this.EVENTS = [];
	}

	/**
	 * Changes the symbol that separates the attributes inside brackets.
	 * @param {string} symbol The new symbol.
	 * @public
	 * @example `
	 *      changeSymbolBetweenAttributes('/')
	 *      => [attr1=et / attr2=voilà]
	 * `
	 */
	public changeSymbolBetweenAttributes(symbol: string): void {
		this.SYMBOL_BETWEEN_ATTRIBUTES = symbol;
	}

	/**
	 * Indents a template in order to concatenate it with another one.
	 * @param {string} template The template to indent.
	 * @param {number} indentation The level of indentation (by default 1).
	 * @returns {string} The new template.
	 * @since 1.0.3
	 */
	public indentTemplate(template: string, indentation: number = 1): string {
		var newTemplate = "";
		var lines = this._extractLinesFrom(template);
		for (var line of lines) {
			newTemplate += ">".repeat(indentation) + line.trim() + "\n"; // \n to add more lines
		}

		return newTemplate.trim();
	}

	/**
	 * Gets the indentation level of a line.
	 * @param {string} line The line to parse.
	 * @return {number} The level of indentation.
	 * @private
	 */
	private _level(line: string): number {
		var level = 0;
		for (var i = 0; i < line.length; i++) {
			if (line[i] !== ">") {
				break;
			} else {
				level++;
			}
		}
		return level;
	}

	/**
	 * Extracts the different lines of a template in order to analyse them individually.
	 * @param {string} template The template of the HTML elements.
	 * @return {Array<string>} The lines from a template.
	 * @private
	 */
	private _extractLinesFrom(template: string): string[] {
		var lines = template.trim().split("\n");
		for (var i = 0; i < lines.length; i++) {
			lines[i] = lines[i].trim();
			if (lines[i].length == 0) {
				lines.splice(i, 1);
			}
		}
		return lines;
	}

	/**
	 * Decodes HTML entities like `&amp;` etc.
	 * @param {string} content The content to decode.
	 * @return {string} The decoded content.
	 * @private
	 * {@link https://stackoverflow.com/questions/7394748/whats-the-right-way-to-decode-a-string-that-has-special-html-entities-in-it/7394787#7394787}
	 */
	private _decodeHTMLEntities(content: string): string {
		var txt = document.createElement("textarea");
		txt.innerHTML = content;
		return txt.value;
	}

	/**
	 * Gets an event according to its name.
	 * @param name The name of the event we are looking for.
	 * @return {{name: string, type: string, callback: Function, options: any}} The event we are looking for.
	 * @private
	 */
	private _searchForEvent(name: string): Listener | null {
		for (var event of this.EVENTS) {
			if (name === event.name) {
				return event;
			}
		}
		return null;
	}

	/**
	 * Generates a new HTML element from a line (you must use a specific syntax & order).
	 * @param {string} line The line to parse.
	 * @return {HTMLElement} The generated HTML element.
	 * @private
	 * @throws If there is no tagname.
	 */
	private _createElementFromLine(line: string): HTMLElement {
		// Be careful when you use exec() with the global flag
		// If you use a global flag, then set the lastIndex property of the regex to 0 (its initial value).
		// this.REGEX.lastIndex = 0;
		//
		// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#finding_successive_matches
		var matches = this.REGEX.exec(line) || [];

		var tagname: string | null = matches[1] || null;
		var classes: string[] | null = matches[2] ? (matches[2].split(".") as string[]).filter((v) => v !== "") : null;
		var id: string | null = matches[3] ? matches[3].replace("#", "") : null;
		var content: string | null = matches[4] || null;
		var attributes: string[] | null = matches[5] ? matches[5].split(this.SYMBOL_BETWEEN_ATTRIBUTES) : null;
		var events: string[] | null = matches[6] ? (matches[6].split(";") as string[]).filter((v) => v !== "") : null;

		if (!tagname) {
			throw new Error('HTMLBuilder: unable to parse a line: "' + line + '"');
		}

		var element: HTMLElement = document.createElement(tagname);
		if (classes) {
			for (var c of classes) {
				if (/\d/.test(c[0])) {
					console.error("HTMLBuilder: invalid syntax for class name '" + c + "'");
					continue;
				}
				element.classList.add(c);
			}
		}
		if (attributes) {
			for (var attr of attributes) {
				if (/\d/.test(attr[0])) {
					console.error("HTMLBuilder: invalid syntax for attribute name '" + attr + "'");
					continue;
				}

				attr = attr.trim();
				if (attr.indexOf("=") !== -1) {
					var name: string = attr.split("=")[0];
					var value: string = attr.split("=")[1];
					element.setAttribute(name, value);
				} else {
					element.setAttribute(attr, "");
				}
			}
		}

		if (id) element.id = id;
		if (content) element.appendChild(document.createTextNode(this._decodeHTMLEntities(content)));

		if (events) {
			for (var name of events) {
				if (/\d/.test(name[0])) {
					console.error("HTMLBuilder: invalid syntax for event name '" + name + "'");
					continue;
				}

				var event: Listener | null = this._searchForEvent(name);
				if (event) {
					// @ts-ignore
					element.addEventListener(event.type, event.callback, event.options);
				}
			}
		}

		return element;
	}

	/**
	 * Gets the maximum level of indentation.
	 * @param {Array} children The list of children of a main element from a template.
	 * @return {number} The maximum level of indentation of a list of children.
	 * @private
	 */
	private _maxLevel(children: [HTMLElement, number][]): number {
		var max: number = children[0][1];
		for (var child of children) {
			var level = child[1];
			if (level > max) {
				max = level;
			}
		}
		return max;
	}

	/**
	 * Gets the index of the deepest element. The deepest element is the last child to have the highest level of indentation.
	 * @param {Array} children The list of children of a main element from a template.
	 * @return {number} The index of the deepest child.
	 * @private
	 */
	private _getIndexOfDeepestElement(children: [HTMLElement, number][]): number {
		var max: number = this._maxLevel(children);
		if (max === 1) {
			// If all the elements are on the closest possible level (1),
			// then we want to append the last child of the list.
			// Remember that we do a prepend() not an append(),
			// therefore the last one must go first in order to keep the right order
			return children.length - 1;
		}

		var lastIndex: number = 1;
		for (var i = 0; i < children.length; i++) {
			var level = children[i][1];
			if (level === max) {
				lastIndex = i;
			}
		}
		return lastIndex;
	}

	/**
	 * Gets the index of the nearest element of the deepest one. This child is the parent element of the deepest one.
	 * @param indexOfDeepest The index of the deepest element.
	 * @param children The list of children of a main element from a template.
	 * @return {number} The index of the nearest child.
	 * @private
	 */
	private _getIndexOfNearestParentElementOf(
		indexOfDeepest: number,
		children: [HTMLElement, number][]
	): number | null {
		var deepest: number = children[indexOfDeepest][1];
		var lastIndex: number | null = null;
		for (var i = 0; i < indexOfDeepest; i++) {
			var level: number = children[i][1];
			if (level === deepest - 1) {
				lastIndex = i;
			}
		}
		return lastIndex;
	}

	/**
	 * Reproduces a template in full HTML structure and adds it to the parent as a child (there can be several children).
	 * @param {string} template The template of your HTML structure.
	 * @public
	 */
	public generate(template: string): void {
		if (template.trim().length === 0) return;

		// We read all the lines in order to identify the main HTML elements,
		// i.e. those without indentation

		var lines: string[] = this._extractLinesFrom(template);
		var mainLines: [string, number][] = [];
		var i = 0;
		var k = 0;

		for (i = 0; i < lines.length; i++) {
			var line = lines[i];
			var level = this._level(line);
			if (level === 0) {
				mainLines.push([line, i]); // the line & its index among all the lines
			}
		}

		// We read the next lines and we create an array [HTMLElement, its level] that we save
		// in a list of children, for each main element.

		for (i = 0; i < mainLines.length; i++) {
			var childrenElements: [HTMLElement, number][] = [];
			var mainLine: string = mainLines[i][0];
			var mainLevel: number = mainLines[i][1];
			var nextMainLevel: number = mainLines[i + 1] ? mainLines[i + 1][1] : lines.length;
			var mainElement: HTMLElement = this._createElementFromLine(mainLine);

			// starts at the position of the main line
			// ends at the position of the next main line
			// in order to get only its children
			for (k = mainLevel + 1; k < nextMainLevel; k++) {
				var line: string = lines[k];
				var child: HTMLElement = this._createElementFromLine(line);
				childrenElements.push([child, this._level(line)]);
			}

			// We search for the deepest element (i.e. the one with the highest level of indentation)
			// This deepest has as parent the nearest element which has a level of indentation equal to "child's level - 1"
			// We call it the "nearest parent element".
			// Then, because we read the list of children from bottom to top, we prepend() in order to keep the right order.
			// Indeed, append() would reverse the right order.

			while (childrenElements.length > 0) {
				var indexOfDeepest: number = this._getIndexOfDeepestElement(childrenElements);
				var indexOfNearestParent: number | null = this._getIndexOfNearestParentElementOf(
					indexOfDeepest,
					childrenElements
				);

				// Don't forget to specify "!== null" because indexOfNearestParent can be 0 (= false)
				indexOfNearestParent !== null
					? childrenElements[indexOfNearestParent][0].prepend(childrenElements[indexOfDeepest][0])
					: mainElement.prepend(childrenElements[indexOfDeepest][0]);

				childrenElements.splice(indexOfDeepest, 1);
			}

			this.parent.appendChild(mainElement);
		}
	}
}

src/HTMLBuilder.js
"use strict";
// FRENCH="àèìòùÀÈÌÒÙáéíóúýÁÉÍÓÚÝâêîôûÂÊÎÔÛãñõÃÑÕäëïöüÿÄËÏÖÜŸçÇßØøÅåÆæœ"
/**
 * A tool that allows you to generate HTML content from a template in an optimised way.
 * @class
 */
class HTMLBuilder {
    /**
     * @constructs HTMLBuilder
     * @param {HTMLElement} parent The parent in which to put the generated elements.
     */
    constructor(parent) {
        /**
         * The regular expression used to parse a template.
         * @type {RegExp}
         * @constant
         * @private
         */
        this.REGEX = /(\w{1,})((?:\.[\w-]*){0,}){0,}(#[\w-]{0,}){0,1}(?:\((.*)\)){0,1}(?:\[(.*)\]){0,1}(?:\@([\w;]*)){0,}/;
        /**
         * The symbol uses to separate different attributes.
         * @type {string}
         * @private
         */
        this.SYMBOL_BETWEEN_ATTRIBUTES = ";";
        /**
         * The list of all the events.
         * @private
         */
        this.EVENTS = [];
        this.parent = parent || document.body;
    }
    /**
     * Changes the parent element.
     *
     * @param parent The new parent element in which to put the generated elements.
     * @public
     */
    setParent(parent) {
        this.parent = parent;
    }
    /**
     * Registers an event to use in a template. Those events are available for all the templates.
     *
     * @param {{name: string, type: string, callback: Function, options: any}} event The event to register.
     * @public
     */
    bindEvent(event) {
        if (!event.name)
            throw new Error("bindEvent(): cannot bind an event without a name.");
        if (!event.type)
            throw new Error("bindEvent(): cannot bind an event without a precise type.");
        if (!event.callback)
            throw new Error("bindEvent(): cannot bind an event without a callback function.");
        if (event.name.startsWith("on")) {
            event.name = event.name.replace("on", "");
        }
        this.EVENTS.push(event);
    }
    /**
     * Changes the symbol that separates the attributes inside brackets.
     *
     * @param {string} symbol The new symbol.
     * @public
     * @example `
     *      changeSymbolBetweenAttributes('/')
     *      => [attr1=et / attr2=voilà]
     * `
     */
    changeSymbolBetweenAttributes(symbol) {
        this.SYMBOL_BETWEEN_ATTRIBUTES = symbol;
    }
    /**
     * Indents a template in order to concatenate it with another one.
     * @param {string} template The template to indent.
     * @param {number} indentation The level of indentation (by default 1).
     * @returns {string} The new template.
     * @since 1.0.3
     */
    indentTemplate(template, indentation = 1) {
        var newTemplate = "";
        var lines = this._extractLinesFrom(template);
        for (var line of lines) {
            newTemplate += ">".repeat(indentation) + line + "\n"; // \n to add more lines
        }
        return newTemplate.trim();
    }
    /**
     * Gets the indentation level of a line.
     *
     * @param {string} line The line to parse.
     * @return {number} The level of indentation.
     * @private
     */
    _level(line) {
        var level = 0;
        for (var i = 0; i < line.length; i++) {
            if (line[i] !== ">") {
                break;
            }
            else {
                level++;
            }
        }
        return level;
    }
    /**
     * Extracts the different lines of a template in order to analyse them individually.
     *
     * @param {string} template The template of the HTML elements.
     * @return {Array<string>} The lines from a template.
     * @private
     */
    _extractLinesFrom(template) {
        var lines = template.trim().split("\n");
        for (var i = 0; i < lines.length; i++) {
            lines[i] = lines[i].trim();
        }
        return lines;
    }
    /**
     * Decodes HTML entities like `&amp;` etc.
     *
     * @param {string} content The content to decode.
     * @return {string} The decoded content.
     * @private
     * {@link https://stackoverflow.com/questions/7394748/whats-the-right-way-to-decode-a-string-that-has-special-html-entities-in-it/7394787#7394787}
     */
    _decodeHTMLEntities(content) {
        var txt = document.createElement("textarea");
        txt.innerHTML = content;
        return txt.value;
    }
    /**
     * Gets an event according to its name.
     *
     * @param name The name of the event we are looking for.
     * @return {{name: string, type: string, callback: Function, options: any}} The event we are looking for.
     * @private
     */
    _searchForEvent(name) {
        for (var event of this.EVENTS) {
            if (name === event.name) {
                return event;
            }
        }
        return null;
    }
    /**
     * Generates a new HTML element from a line (you must use a specific syntax & order).
     *
     * @param {string} line The line to parse.
     * @return {HTMLElement} The generated HTML element.
     * @private
     * @throws If there is no tagname.
     */
    _createElementFromLine(line) {
        // Be careful when you use exec() with the global flag
        // If you use a global flag, then set the lastIndex property of the regex to 0 (its initial value).
        // this.REGEX.lastIndex = 0;
        //
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#finding_successive_matches
        var matches = this.REGEX.exec(line) || [];
        var tagname = matches[1] || null;
        var classes = matches[2] ? matches[2].split(".").filter((v) => v !== "") : null;
        var id = matches[3] ? matches[3].replace("#", "") : null;
        var content = matches[4] || null;
        var attributes = matches[5] ? matches[5].split(this.SYMBOL_BETWEEN_ATTRIBUTES) : null;
        var events = matches[6] ? matches[6].split(";").filter((v) => v !== "") : null;
        if (!tagname) {
            throw new Error('HTMLBuilder: unable to parse a line: "' + line + '"');
        }
        var element = document.createElement(tagname);
        if (classes) {
            for (var c of classes) {
                if (/\d/.test(c[0])) {
                    console.error("HTMLBuilder: invalid syntax for class name '" + c + "'");
                    continue;
                }
                element.classList.add(c);
            }
        }
        if (attributes) {
            for (var attr of attributes) {
                if (/\d/.test(attr[0])) {
                    console.error("HTMLBuilder: invalid syntax for attribute name '" + attr + "'");
                    continue;
                }
                attr = attr.trim();
                if (attr.indexOf("=") !== -1) {
                    var name = attr.split("=")[0];
                    var value = attr.split("=")[1];
                    element.setAttribute(name, value);
                }
                else {
                    element.setAttribute(attr, "");
                }
            }
        }
        if (id)
            element.id = id;
        if (content)
            element.appendChild(document.createTextNode(this._decodeHTMLEntities(content)));
        if (events) {
            for (var name of events) {
                if (/\d/.test(name[0])) {
                    console.error("HTMLBuilder: invalid syntax for event name '" + name + "'");
                    continue;
                }
                var event = this._searchForEvent(name);
                if (event) {
                    // @ts-ignore
                    element.addEventListener(event.type, event.callback, event.options);
                }
            }
        }
        return element;
    }
    /**
     * Gets the maximum level of indentation.
     *
     * @param {Array} children The list of children of a main element from a template.
     * @return {number} The maximum level of indentation of a list of children.
     * @private
     */
    _maxLevel(children) {
        var max = children[0][1];
        for (var child of children) {
            var level = child[1];
            if (level > max) {
                max = level;
            }
        }
        return max;
    }
    /**
     * Gets the index of the deepest element. The deepest element is the last child to have the highest level of indentation.
     *
     * @param {Array} children The list of children of a main element from a template.
     * @return {number} The index of the deepest child.
     * @private
     */
    _getIndexOfDeepestElement(children) {
        var max = this._maxLevel(children);
        if (max === 1) {
            // If all the elements are on the closest possible level (1),
            // then we want to append the last child of the list.
            // Remember that we do a prepend() not an append(),
            // therefore the last one must go first in order to keep the right order
            return children.length - 1;
        }
        var lastIndex = 1;
        for (var i = 0; i < children.length; i++) {
            var level = children[i][1];
            if (level === max) {
                lastIndex = i;
            }
        }
        return lastIndex;
    }
    /**
     * Gets the index of the nearest element of the deepest one. This child is the parent element of the deepest one.
     *
     * @param indexOfDeepest The index of the deepest element.
     * @param children The list of children of a main element from a template.
     * @return {number} The index of the nearest child.
     * @private
     */
    _getIndexOfNearestParentElementOf(indexOfDeepest, children) {
        var deepest = children[indexOfDeepest][1];
        var lastIndex = null;
        for (var i = 0; i < indexOfDeepest; i++) {
            var level = children[i][1];
            if (level === deepest - 1) {
                lastIndex = i;
            }
        }
        return lastIndex;
    }
    /**
     * Reproduces a template in full HTML structure and adds it to the parent as a child (there can be several children).
     *
     * @param {string} template The template of your HTML structure.
     * @public
     */
    generate(template) {
        if (template.trim().length === 0)
            return;
        // We read all the lines in order to identify the main HTML elements,
        // i.e. those without indentation
        var lines = this._extractLinesFrom(template);
        var mainLines = [];
        var i = 0;
        var k = 0;
        for (i = 0; i < lines.length; i++) {
            var line = lines[i];
            var level = this._level(line);
            if (level === 0) {
                mainLines.push([line, i]); // the line & its index among all the lines
            }
        }
        // We read the next lines and we create an array [HTMLElement, its level] that we save
        // in a list of children, for each main element.
        for (i = 0; i < mainLines.length; i++) {
            var childrenElements = [];
            var mainLine = mainLines[i][0];
            var mainLevel = mainLines[i][1];
            var nextMainLevel = mainLines[i + 1] ? mainLines[i + 1][1] : lines.length;
            var mainElement = this._createElementFromLine(mainLine);
            // starts at the position of the main line
            // ends at the position of the next main line
            // in order to get only its children
            for (k = mainLevel + 1; k < nextMainLevel; k++) {
                var line = lines[k];
                var child = this._createElementFromLine(line);
                childrenElements.push([child, this._level(line)]);
            }
            // We search for the deepest element (i.e. the one with the highest level of indentation)
            // This deepest has as parent the nearest element which has a level of indentation equal to "child's level - 1"
            // We call it the "nearest parent element".
            // Then, because we read the list of children from bottom to top, we prepend() in order to keep the right order.
            // Indeed, append() would reverse the right order.
            while (childrenElements.length > 0) {
                var indexOfDeepest = this._getIndexOfDeepestElement(childrenElements);
                var indexOfNearestParent = this._getIndexOfNearestParentElementOf(indexOfDeepest, childrenElements);
                // Don't forget to specify "!== null" because indexOfNearestParent can be 0 (= false)
                indexOfNearestParent !== null
                    ? childrenElements[indexOfNearestParent][0].prepend(childrenElements[indexOfDeepest][0])
                    : mainElement.prepend(childrenElements[indexOfDeepest][0]);
                childrenElements.splice(indexOfDeepest, 1);
            }
            this.parent.appendChild(mainElement);
        }
    }
}
